### DISCUSSION TITLE
Restore violentmonkey scripts after removing extension?

### DISCUSSION THREAD

**Discussion by omphp** [violentmonkey/violentmonkey#1674] [Q&A]
I accidentally removed my violentmonkey extension recently without backing up my scripts. Is there any way to restore them?

side note: I followed this guide (https://superuser.com/questions/1149988/how-can-i-restore-a-lost-tampermonkey-userscripts) and could find parts of my script in 000003.log file but it all garbled and can't make sense of it. Any tips?

**Comment by oooooooo** [comment]
If you still have the *.ldb, you can restore it by.

```python
# pip install leveldb
#
# Windows usage: python main.py "C:\Users\<USER>\AppData\Local\<VENDOR>\<BROWSER>\User Data\Default\Local Extension Settings\<EXTENSION_ID>"
#          i.e.: python main.py "C:\Users\foo\AppData\Local\Google\Chrome\User Data\Default\Local Extension Settings\jinjaccalgkegednnccohejagnlnfdag
#   Linux usage: python main.py "/home/<USER>/.config/<BROWSER>/Default/Local Extension Settings/<EXTENSION_ID>"
#          i.e.: python main.py "/home/foo/.config/google-chrome-beta/Default/Local Extension Settings/jinjaccalgkegednnccohejagnlnfdag
#     Mac usage: python main.py "/Users/<USER>/Library/Application Support/Google/Chrome/Default/Local Extension Settings/<EXTENSION_ID>/"
#          i.e.: python main.py "/Users/foo/Library/Application Support/Google/Chrome/Default/Local Extension Settings/jinjaccalgkegednnccohejagnlnfdag

import codecs
import json
import leveldb
import re
import sys

pattern = re.compile("code:\d+")

db = leveldb.LevelDB(sys.argv[1:][0])
for k, v in db.RangeIter():
    key = k.decode("utf-8")
    if not pattern.match(key):
        continue

    code = json.JSONDecoder().decode(v.decode("utf-8"))

    filename = key.replace(":", "") + ".js"
    with codecs.open(filename, "w", "utf-8") as f:
        f.write(code)
```

**Reply by rudolphos** [reply]
I tried this script on these files:
![image](https://github.com/user-attachments/assets/7407745a-95a2-48dd-a569-35785e4d2c88)
, but getting an error:
```
lubuntu@lubuntu:~/Desktop$ python3 script.py ~/Desktop/jinjaccalgkegednnccohejagnlnfdag/
Traceback (most recent call last):
  File "/home/lubuntu/Desktop/script.py", line 18, in <module>
    db = leveldb.LevelDB(sys.argv[1:][0])
leveldb.LevelDBError: Corruption: missing start of fragmented record(2)
```
This db I'm trying to recover is taken from Windows shadow copy.
Is it possible that Windows use some compression algorithm on files without file extensions (like Manifest "File")? Years ago I also tried restoring sqlite db files for sessionbuddy extension and they were always corrupt for some reason, and they never had a file extension.
When I open ldb files with notepad, it looks like the scripts are there but among various other characters. I haven't found any tools that could dump such db with idb comparator.

**Comment by jesus2099** [comment]
This Windows path converts to:

> `%LocalAppData%\Vivaldi\User Data\Default\Local Extension Settings\jinjaccalgkegednnccohejagnlnfdag`

For everyone, paste this in your File Explorer and you will get there.
But maybe it's no use for the Python script.

I need to restore a userscript I was developing but got erase by auto-update (IO forgot to set a future date, like I usually do, or to use the disallow update, that I never tried yet).
I will try if I can retrieve older version with this. 🤞 

Thank you, @oooooooo.

Update:
On Windows `pip install leveldb` does not work:

> _Don't know how to compile leveldb for Windows!_

So it worked on Linux and your script ran very fine.

But unfortunately for my specific need, Violentmonkey does not store older versions of scripts, apparently.

**Comment by rudolphos** [comment]
I'm in a similar situation. Violentmonkey extension crashed for some reason on the Brave browser and I accidentally clicked on the "Repair" option. Good thing I have shadow copy enabled, so I have access to the same indexeddb and Local Extension Settings folders from previous days. The problem is that I can't restore them on Windows in any way due to no leveldb packages available via pip for Windows.

```
Collecting leveldb
  Using cached leveldb-0.201.tar.gz (236 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error

  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [1 lines of output]
      Don't know how to compile leveldb for Windows!
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

```

Also, there is a tool [LevelDBDumper](https://github.com/mdawsonuk/LevelDBDumper), but it doesn't support `idb_cmp1` comparator used in Chromium browsers.

**Reply by tophf** [reply]
I usually install the extension, exit the browser, go to the new `Local Extension Settings`, remove the new `jinjaccalgkegednnccohejagnlnfdag`, copy the old `jinjaccalgkegednnccohejagnlnfdag` here, then start the browser.

**Reply by rudolphos** [reply]
Did that.

1. Removed existing VM install
2. Installed VM again, quit the browser
4. Used Nirsoft ShadowCopyView "Copy selected files to"
5. Checked. Folder exists with ldb files each around 1-4 MB, also MANIFEST file is there.
6. Start the browser. ViolentMonkey is enabled now
8. Go to ViolentMonkey and it's all empty "No script is found."
![image](https://github.com/user-attachments/assets/750e88ed-7411-4b7e-b4dc-1e6243e54367)

**Reply by tophf** [reply]
Try an older backup. Also, maybe some sub-process of the browser remained in memory? I use the `Exit` command in the browser menu.

**Reply by rudolphos** [reply]
I made sure all browser processes are ended in the task manager.
Restored an older backup (edit: tried already 2 older backups). When opened, VM appears to be damaged (faulty drop downs, missing values) for some reason
![31 Violentmonkey - Brave](https://github.com/user-attachments/assets/4cd87eb2-e066-4e2f-858b-f8aed9fcb25f)
![24 Violentmonkey - Brave](https://github.com/user-attachments/assets/90425345-cbe4-486f-be10-422bbd3b92b9)

In VM console it shows this:
```
index.js:1 Uncaught (in promise) Error: Invalid JSON
    at index.js:1:1429
    at index.js:1:137897
    at index.js:1:139876
    at index.js:1:194219
```

Once disabled/enabled, VM seems to be completely reset and in the basic blank state. All the Local Extension Settings files I restored were automatically moved to `/lost` folder.
VM console output:
`Upgrade database... index.js:1 `

**Comment by Pixopax** [comment]
As someone seeks a way to get back the scripts after Google decided to kill tampermonkey extension from chrome:
Locate the .log file from C:\Users\<username>\AppData\Local\Google\Chrome\User Data\Default\Local Extension Settings\jinjaccalgkegednnccohejagnlnfdag

It contains all scripts, but is not usable directly.
It contains also several versions of your scripts with the last version on the end of the file. 
Open it in a editor like Notepad++, read it, locate the parts where your script is, and copy it from the file, it usually starts with 
// ==UserScript== and ends with ();\n

Then use chatgpt to clean it up, it does perfectly.
Just ask it to extract the violentmonkey script from a .log file out of chrome. 
Then paste the text part. Et voila, you have your script back.

Then you can copy the code into a new script in firefox tampermonkey extension.

**Comment by KPG-Kane** [comment]
Chrome updated and took ViolentMonkey access away from me, so I could not access my script source code.

I created a Python tool to export it from the Chrome profile data. Thought it could be useful to others, so I published it here:

[ViolentMonkey  Script Restorer](https://github.com/KPG-Kane/violent-monkey-script-restorer)

**Reply by drsweb** [reply]
> Chrome updated and took ViolentMonkey access away from me, so I could not access my script source code.
> 
> I created a Python tool to export it from the Chrome profile data. Thought it could be useful to others, so I published it here:
> 
> [ViolentMonkey Script Restorer](https://github.com/KPG-Kane/violent-monkey-script-restorer)

Big winner here.  Thank you very much for sharing.  Love ViolentMonkey RIP, don't love the decision to store the scripts in this format.

**Reply by YutMarma** [reply]
> Chrome updated and took ViolentMonkey access away from me, so I could not access my script source code.
> 
> I created a Python tool to export it from the Chrome profile data. Thought it could be useful to others, so I published it here:
> 
> [ViolentMonkey Script Restorer](https://github.com/KPG-Kane/violent-monkey-script-restorer)

You, sir, are a gentleman, a scholar, and, what's more, an a-okay chap.

**Comment by skygate2012** [comment]
@KPG-Kane I tried this but only recovered two scripts. Apparently not all my scripts are in the WAL log.
In the end I just downloaded an older version of Chrome (from [ungoogled-chromium](https://github.com/ungoogled-software/ungoogled-chromium)), loaded the unpacked extension there, replaced all the files under `Local Extension Settings`, and exported all my scripts within Violentmonkey.

**Comment by seifer7** [comment]
I wrote a Go script that will export all your scripts from the database file, not parsing log files which can cause missing entries.

I've released the source on Github, and also ready-to-run compiled versions for Windows or Linux.
https://github.com/seifer7/violentmonkey_script_recover/releases

It's dead simple to use.
Run the executable, press S (search), press A (export All).

It has other options too for more advanced use-cases.
